README
radstock

Radstock is primarily a sysadmins tool to diagnose problems with
radius server configurations. It was written because I always found
snoop/tcpdump to be a poor alternative, and trying to use these on
heavily loaded, complex radius servers proved to be a nightmare.

It provides the ability to fully decode each packet. Here is some
sample output. 

unix# radstock
 
interface: hme0 (1.2.3.0/255.255.255.0)     filter: udp and port 1645
Auth-Request(134) - 10.1.3.6:60284 -> 1.2.3.4:1645              
  User-Name              Len 29         "bobby@dummy.org"
  CHAP-Password          Len 19         "*****`*p*M|**l**N"
  NAS-IP-Address         Len  6         10.1.3.6
  NAS-Port-Id            Len  6         117
  Service-Type           Len  6         Framed-User
  Framed-Protocol        Len  6         PPP
  Called-Station-Id      Len  8         "9876"
  Calling-Station-Id     Len 12         "12345678"
  NAS-Port-Type          Len  6         Async
Auth-Reject (35) - 1.2.3.4:1645 -> 10.1.3.8:60284              
Auth-Request(113) - 10.1.1.1:36332 -> 1.2.3.4:1645              
  User-Name              Len  8         "paul"
  CHAP-Password          Len 19         "****Y***W*>N****d"
  NAS-IP-Address         Len  6         10.1.1.1
  NAS-Port-Id            Len  6         93
  Service-Type           Len  6         Framed-User
  Framed-Protocol        Len  6         PPP
  Called-Station-Id      Len  8         "9876"
  Calling-Station-Id     Len 12         "12345679"
  NAS-Port-Type          Len  6         Async
Auth-Accept (59) - 1.2.3.4:1645 -> 10.1.1.1:36332              
  Service-Type           Len  6         Framed-User
  Framed-Protocol        Len  6         PPP
  Framed-IP-Address      Len  6         255.255.255.254
  Ascend-Idle-Limit      Len  6         1200

However, the key feature of radstock is its ability to filter the
packets shown based on any attribute. It will also listen out for
responses to matched packets and display these as well. An example
filter would be (all on one line):

   radstock -e "(user-name = paul or user-name = bob) and 
   exists nas-port-id"

You can also get radstock to read filters from the command line or
stdin. For further details on how to write a radius filter, see the
manual page.

For those of you who are interested as to the name - here's a
clue. This program was written to replace a program called radsnoop.

COMPILING

To compile radstock you need the following

  o  libpcap
  o  flex or lex
  o  bison/yacc

Once you have these, the following three commands should do just about
all you need.

	./configure 
	make 
	make install

It has been successfully compiled on Linux and Solaris
platforms. Whether it works on them is a completely different story.





ChangeLog
2015-06-18  Eddie Stassen <estassen@gmail.com>

	* Instead of just picking a first available interface as default, rather look for
	one with a vailid looking IP.  Only been tested on Centos 7

2013-11-18  Eddie Stassen <estassen@gmail.com>

	* Added support for Dis-Req, Dis-ACK and Dis-NAK packet types
	
	* Fixed all the buffer overflow issues with output buffer and large 
	  accounting packets
	 
	* Updated to support flex
	
	* Added -S command line option to allow displaying invalid packets. 
	  By default invalid packets will be suppressed

	* Better display for VSA's.  VSA's are not fully supported (this would
	require major changes), but at least they are displayed in a more or
	less readable format.

	* Timestamps for the -T option are formatted better

2000-10-26  Paul Rhodes  <withnail@users.sourceforge.net>

	* radstock.c: Actually uses the value passed on the command line

	* Makefile.in: Passes the dictionary as a define on the command line

	* radstock.1: Added notes to clarify the logic with !=. != will
	not match packets without that attribute at all.

	* process.c: parse_packet should never return FIL_UNDEF for any
	terms. It should return TRUE on UNDEF and the operator is nexists
	and FALSE otherwise.

	* grammar.y: By sending NULL as the last parameter for exists
	operators, we caused an exception when debugging was switched on.



init.c
/*
 * $Id: init.c,v 1.1.1.1 2000/10/13 12:23:11 withnail Exp $
 *
 * Please refer to the COPYRIGHT file for more information. 
 * 
 */

#if defined(BSD) || defined(SOLARIS)
#include <unistd.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <net/if.h>
#include <regex.h>
#include <fcntl.h>
#endif

#if defined(OSF1)
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <net/route.h>
#include <sys/mbuf.h>
#endif

#if defined(LINUX)
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <arpa/inet.h>
#include <ctype.h>
#include <regex.h>
#endif

#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>

#include <pcap.h>
#include <pcap/bpf.h>

#include <stdlib.h>
#include <string.h>
#include <signal.h>

#include "radstock.h"
#include "y.tab.h"
static char rcsid[] = "$Id: init.c,v 1.1.1.1 2000/10/13 12:23:11 withnail Exp $";

int t_radvals = 0;
int t_radattrs = 0;
int curnode = 0;

#define ATTR_NOTFOUND 0
#define ATTR_MATCH    1
#define ATTR_NOMATCH  2

#ifdef YYDEBUG
extern int yydebug;
#endif
char yacc_input_buff[1000];
char *yacc_input_pos;
int yyparse(void);



struct attr_listelem
{
  int attr_index;
  int attr_op;
  union {
    int ivalue;
    regex_t *rvalue;
  } data;
  int result;			/* ATTR_NOTFOUND etc */

};

struct attr_list
{
  struct attr_listelem elem[50];
  int t_elem;
};

struct s_tree tree;

int my_find_radattr(char *attrname, int *radindex)
{
  int lp;

  for (lp = 0; lp < t_radattrs; lp++)
  {
    if (strcasecmp(attrname, radattrs[lp].name) == 0) {
      *radindex = lp;
      return radattrs[lp].vala;
    }
  }
  return 0;
}

int my_find_radval(int radattr, char *attrname)
{
  int lp, tmpval;

  for (lp = 0; lp < t_radvals; lp++)
  {
    if (strcasecmp(attrname, radvals[lp].name) == 0) {
      tmpval = radvals[lp].vala >> 8;
      if (tmpval == radattr)
	
	return radvals[lp].vala - (tmpval << 8);
    }
  }
  return 0;
}


int my_find_attrs(const void *a, const void *b)
{
  struct attribs b1;
  unsigned int a1;

  a1 = (unsigned int) *((int *)a);

  b1 = *((struct attribs *) b);
  //  printf("Testing value %d against attribute table entry %d\n",
  //	 a1, b1.vala);
  
  return (a1 < b1.vala ? -1 : a1 > b1.vala ? 1 : 0);
}

struct attribs *find_radattr_by_value(int lookupval)
{
  return bsearch(&lookupval, 
		 radattrs, t_radattrs, 
		 sizeof(struct attribs), my_find_attrs);
}

struct attribs *find_radvals_by_value(int attrib, int lookupval)
{
  int tmp = (attrib << 8) +lookupval;
  return bsearch(&tmp, 
		 radvals, t_radvals, 
		 sizeof(struct attribs), my_find_attrs);
}


int cmp_attrs(const void *a, const void *b)
{
  struct attribs a1;
  struct attribs b1;

  b1 = *((struct attribs *) b);
  a1 = *((struct attribs *) a);

  return (a1.vala < b1.vala ? -1 : a1.vala > b1.vala ? 1 : 0);
}



int readdict(char *dictfile)
{
  char buffer[BUFSIZE+5];
  char arg1[BUFSIZE], arg2[BUFSIZE], arg3[BUFSIZE], arg4[BUFSIZE];
  int temp2 = 0;
  int lp;

  FILE *fd;

  if ((fd = fopen(dictfile, "r")) == NULL)
  {
    return 0;
  }
  t_radattrs = 0;
  t_radvals = 0;

  // Search for attributes
  while (feof(fd) == 0)
  {
    fgets(buffer, BUFSIZE, fd);
    if (!buffer || *buffer == '#' || *buffer == ' ' || *buffer == '\n' ) continue;
    if (debug & 2) fprintf(stderr, "readdict: buffer = :%s:\n", buffer);
    sscanf(buffer,"%s %s %s %s\n", arg1, arg2, arg3, arg4);

    // Ok - lets check for type of attrib
    if (strcasecmp(arg1, "ATTRIBUTE") == 0)
    {
      // Lets skip this attribute if we have read in our maximum
      if (t_radattrs == (ATTRSIZE - 2)) continue;
      radattrs[t_radattrs].vala = atoi(arg3);
      strcpy(radattrs[t_radattrs].name, arg2);
      if (strcasecmp(arg4, "integer") == 0) temp2 = RADAT_INT;
      else if (strcasecmp(arg4, "ipaddr") == 0) temp2 = RADAT_IP;
      else if (strcasecmp(arg4, "string") == 0) temp2 = RADAT_STR;
      if (debug & 2) 
	fprintf(stderr, "readdict: new attrib (%lu) = :%s:\n",
		radattrs[t_radattrs].vala,
		radattrs[t_radattrs].name);
      radattrs[t_radattrs++].valb = temp2;
    }
  }

  qsort((void *) radattrs, t_radattrs, sizeof(struct attribs), cmp_attrs);
  rewind(fd);

  while (feof(fd) == 0)
  {
    fgets(buffer, BUFSIZE, fd);
    if (!buffer || *buffer == '#' || *buffer == ' ' || *buffer == '\n' ) continue;
    //    printf("Buffer = :%s:\n", buffer);
    sscanf(buffer,"%s %s %s %s\n", arg1, arg2, arg3, arg4);


    // Ok - lets check for type of attrib
    if (strcasecmp(arg1, "VALUE") == 0)
    {

      // Lets skip this attribute if we have read in our maximum
      if (t_radvals == (ATTRSIZE - 2)) continue;

      // Search thru the whole list since it isnt sorted on name
      for (lp = 0; lp < t_radattrs; lp++)
	if (strcasecmp(radattrs[lp].name, arg2) == 0)
	{
	  radvals[t_radvals].vala = (radattrs[lp].vala << 8) + atoi(arg4);
	  strcpy(radvals[t_radvals].name, arg3);
	  t_radvals++;
	  break;
	}
    }
  }
  qsort((void *) radvals, t_radvals, sizeof(struct attribs), cmp_attrs);
  fclose(fd);
  return 1;
}


void get_radfilter(char *arg) 
{
  strcpy(yacc_input_buff, arg);
  yacc_input_pos = yacc_input_buff;

  if (debug) printf("get_radfilter: Parsing radius filter\n");
#ifdef YYDEBUG
  yydebug = 1;
#endif
  yyparse();
  tree.root = curnode - 1;
  if (debug) printf("get_radfilter: Passed radius filter\n");

}


#define FIND_ATTR 1
#define FIND_EQ   2
#define FIND_VAL  3

#if 0

void get_radfilter(char *arg) 
{
  char *achr, *bchr;
  char tmpstr[40], tmpstr2[50], tmpptr;
  int match;
  int radattr = 0, radval, attrindex;

  if (debug) printf("Parsing radius filter\n");
  //  if (!*arg)
  //  return NULL;

  t_radfilters = 0;
  match = FIND_ATTR;

  achr = arg;
    
    while (*achr) {
      if (debug) printf("here we go again\n");
      bchr = tmpstr;
      /* Remove leading space */
      while (isspace(*achr)) achr++;
      while ((*achr) && (*achr != '=') && !isspace(*achr) )
      {
	//printf("achr %c \n", *achr);
	*bchr++ = *achr++;
      }
      /* Remove trailing space */
      while (isspace(*achr)) achr++;
      *bchr = '\0';
      if (debug) printf("found string '%s' (achr '%s')\n", 
			tmpstr, achr);

      switch (match) {
      case FIND_ATTR:
	if ((radattr = my_find_radattr(tmpstr, &attrindex)) == 0) {
	  if (debug) printf("Attribute '%s' not known - exiting\n", tmpstr);
	  clean_exit(-1);
	}
	if (debug) printf("Found attrib (attrnum %d - index %d) %s\n", 
			  radattr, attrindex, tmpstr);
	match = FIND_EQ;
	break;
      case FIND_EQ:
	if (*achr == '=')
	{
	  if (debug) printf ("Found EQ\n");
	  achr++;
	  match = FIND_VAL;
	}
	else {
	  if (debug) printf("Equals not found\n");
	  exit(0);
	}
	break;
      case FIND_VAL:
	// Ok - now we need to know what type of attribute
	// we are looking at
	switch (radattrs[attrindex].valb) {
	case RADAT_INT:
	  if ((radval = my_find_radval(radattr, tmpstr)) == 0)
	    radval = atoi(tmpstr);
	  if (debug) printf("Found enum value (valnum %d) %s\n", radval, tmpstr);
	  filters[t_radfilters].vala = radattr;
	  filters[t_radfilters].valb = radval;
	  t_radfilters++;
	  break;

	case RADAT_IP:
	case RADAT_STR:
	  if (debug && radattrs[attrindex].valb == RADAT_STR) 
	    fprintf(stderr, "Found str value %s\n", tmpstr);
	  if (debug && radattrs[attrindex].valb == RADAT_IP) 
	    fprintf(stderr, "Found IP value %s\n", tmpstr);
	  
	  *tmpstr2 = '\0';
	  tmpptr = tmpstr;
	  if (*tmpptr == '\'') tmpptr++;
	  fprintf(stderr, "trailing '%c'\n", tmpstr[strlen(tmpstr)-1]);
	  
	  if (tmpstr[strlen(tmpstr)-1] == '\'')
	  {
	    fprintf(stderr, "trailing '\n");
	  }
	  if (*tmpptr != '^') strcat(tmpstr2, "^");
	  strcat(tmpstr2, tmpptr);
	  
	  if (*(tmpstr+strlen(tmpstr)-1) != '$') strcat(tmpstr2, "$");
	  if (debug) fprintf(stderr, "Adding regexp '%s'\n", tmpstr2);

	  filters[t_radfilters].vala = radattr;
	  filters[t_radfilters].valb = (int) malloc(sizeof(regex_t));
	  if (filters[t_radfilters].valb == (int) NULL) clean_exit(-2);
	  regcomp((regex_t *) filters[t_radfilters].valb,
		  tmpstr2, REG_EXTENDED|REG_NOSUB|REG_ICASE);
	  //strcpy(filters[t_radfilters].name, tmpstr);
	  t_radfilters++;
	  break;
	}
	match = FIND_ATTR;
	
	break;
      }
    }
  
}

#endif


int yacc_find_attr(char *str)
{
  int index, res;
  res = my_find_radattr(str,&index);
  if (res == 0)
  {
    fprintf(stderr," Attribute '%s' not found\n", str);
    clean_exit(-2);
  }
  //if (debug) fprintf(stderr, "Found attribute (%d) '%s'\n", 
  //res, str);
  return res;
}


int yacc_parse_test(int a1, int a2, int a3, char *value)
{
  struct attribs *myattr;
  int retval;

  myattr = find_radattr_by_value(a1);
  if (debug) 
    fprintf(stderr, "*** Test: attr (index %d) %s  op %d  value (type %d) '%s'\n",
	    a1, myattr->name, a2, a3, value);
  
  /*
   * What we do now is purely decided on the type of attribute
   */
  if (a2 == LEX_EXIST || a2 == LEX_NEXIST)
  {
    filters[t_radfilters].vala = a1;
    filters[t_radfilters].op   = a2;
    filters[t_radfilters].valb = (unsigned long)NULL;
    retval = t_radfilters++;
  }
  else 
  switch (myattr->valb)
  {
  case RADAT_INT:
    {
      int radval;

      if (a3 == LEX_NUM) 
      {
	radval = atoi(value);
      }
      else if ((radval = my_find_radval(a1, value)) == 0)
      {
	fprintf(stderr, "Error: Attribute %s - no value defined for %s\n",
		myattr->name, value);
	clean_exit(-2);
      }
      if (debug) fprintf(stderr, "Found integer %d\n", radval);
      filters[t_radfilters].vala = a1;
      filters[t_radfilters].op   = a2;
      filters[t_radfilters].valb = radval;
      retval = t_radfilters++;
    }
    break;
  case RADAT_IP:
  case RADAT_STR:
    {
      char tmpstr[50], *tmpptr;
      if (a2 != LEX_EQ && a2 != LEX_NEQ) 
      {
	fprintf(stderr, "Only = and != are valid for STR and IP types\n");
	clean_exit(-2);
      }
      *tmpstr = '\0';
      if (*value != '^') strcat(tmpstr, "^");

      tmpptr = value;
      if (*tmpptr == '\'') tmpptr++;
	  
      if (value[strlen(value)-1] == '\'')
	value[strlen(value)-1] = '\0';

      strcat(tmpstr, tmpptr);
      if (*(value+strlen(value)-1) != '$') strcat(tmpstr, "$");
      if (debug) fprintf(stderr, "Adding regexp '%s'\n", tmpstr);

      filters[t_radfilters].vala = a1;
      filters[t_radfilters].op   = a2;
      filters[t_radfilters].valbp = malloc(sizeof(regex_t));
      if (filters[t_radfilters].valbp ==  NULL) clean_exit(-2);
      regcomp((regex_t *) filters[t_radfilters].valbp,
	      tmpstr, REG_EXTENDED|REG_NOSUB|REG_ICASE);
      //strcpy(filters[t_radfilters].name, tmpstr);
      retval = t_radfilters++;
    }
    break;
  }
  return retval;
}


int yacc_add_node(int a1, int a2, int a3)
{
  int retval;

  retval = curnode++;
  if (debug) 
    fprintf(stderr, "Node %d: left %d  op %d  right %d\n",
	    retval, a1, a2, a3);
  tree.elem[retval].operator  = a2;
  tree.elem[retval].lside     = a1;
  tree.elem[retval].rside     = a3;

  return retval;
}


radstock.c
/*
 * $Id: radstock.c,v 1.8 2000/10/26 15:16:42 withnail Exp $
 *
 * Please refer to the COPYRIGHT file for more information. 
 * 
 */

#if defined(BSD) || defined(SOLARIS)
#include <unistd.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <regex.h>
#include <fcntl.h>
#endif

#if defined(OSF1)
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <net/route.h>
#include <sys/mbuf.h>
#endif

#if defined(LINUX)
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <arpa/inet.h>
#include <ctype.h>
#include <regex.h>
#include <time.h>
#endif


#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>

#include <pcap.h>
#include <pcap/bpf.h>

#include <stdlib.h>
#include <string.h>
#include <signal.h>

#include "radstock.h"
//#include "version.h"

static char rcsid[] = 
  "$Id: radstock.c,v 1.8 2000/10/26 15:16:42 withnail Exp $";


//#define DICTFILE "/usr/local/etc/raddb/dictionary"

char defpcapfilter[] = "udp and port 1645";

int snaplen = 65535, promisc = 1, to = 1000;
int show_empty = 0, show_hex = 0;
int match_after = 0, keep_matching = 0;
int invert_match = 0, bin_match = 0;
int matches = 0, max_matches = 0;
int live_read = 1, want_delay = 0;;

char pc_err[PCAP_ERRBUF_SIZE], *re_err;

int re_match_word = 0, re_ignore_case = 0;

char *match_data = NULL, *bin_data = NULL, *filter = NULL;
int match_len = 0;
int debug = 0, quiet = 0;
struct bpf_program pcapfilter;
struct in_addr net, mask;
pcap_t *pd = NULL;
char *dev = NULL;
int link_offset;

int show_auth = 1;
int show_acct = 1;
int show_req  = 1;
int show_resp = 1;
int show_invalid = 0;

char *read_file = NULL, *dump_file = NULL, *dict_file = NULL;
pcap_dumper_t *pd_dump = NULL;

struct timeval prev_ts = {0, 0}, prev_delay_ts = {0,0};
char *(*print_time_fn)() = NULL;



int main(int argc, char **argv) {
  int c;
  char *radfilterfile = NULL;
  char radfilterbuf[1000] = "\0";
  pcap_if_t *alldevs;
  pcap_if_t *d;
  bpf_u_int32 netp,maskp;

  signal(SIGINT,  clean_exit);
  signal(SIGQUIT, clean_exit);
  signal(SIGABRT, clean_exit);
  signal(SIGPIPE, clean_exit);


  debug = 0; quiet = 0;
  /* 
   * Lets allocate a temporary buffer to store the expression in.
   * Then, we can fill this buffer from whatever means we like
   * STDIN, file, or optarg, and treat it exactly the same 
   */
  if (debug) printf("Parsing switches\n");
  while ((c = getopt(argc, argv, "KLMNpd:hXVi:c:wqe:vxlDtSTn:i:A:I:O:f:")) != EOF) {
    switch (c) {
    case 'K':
      if (show_auth == 1) show_acct = 0;
      break;
    case 'L':
      if (show_acct == 1) show_auth = 0;
      break;
    case 'M':
      if (show_req == 1) show_resp = 0;
      break;
    case 'N':
      if (show_resp == 1) show_req = 0;
      break;
    case 'd':
      if (!isdigit(*optarg))
      {
	printf("error: switch -d requires an integer value\n");
	clean_exit(-4);
      }
      debug = atoi(optarg);
      break;
    case 'p':
      promisc = 0;
    case 'f':
      dict_file = optarg;
      break;
    case 'I':  
      read_file = optarg;
      break;
    case 'O':
      dump_file = optarg;
      break;
    case 'A': 
      match_after = atoi(optarg) + 1;
      break;
    case 'i': 
      dev = optarg;
      break;
    case 'n':
      max_matches = atoi(optarg);
      break;
    case 'T':
      print_time_fn = &print_time;
      break;
    case 'D':
      print_time_fn = &print_time_date;
      break;
    case 'c':
      radfilterfile = optarg;
      break;
    case 't':
      //print_time = &print_time_absolute;
      break;
    case 'l':
      setvbuf(stdout, NULL, _IOLBF, 0);
      break;
    case 'x':
      show_hex++;
      break;
    case 'v':
      invert_match++;
      break;
    case 'e':
      strcpy(radfilterbuf,optarg);
      /* Used to be --- show_empty++; */
      break;
    case 'q':
      quiet++;
      break;
    case 'S':
      show_invalid = 1;
      break;
    case 'w':
      re_match_word++;
      break;
    case 'V':
      version();
    case 'X':
      bin_match++;
      break;
    case 'h':
      usage(0);
    }
  }
  if (debug) printf("Reading dictionary\n");

  if (dict_file) 
  {
    if (readdict(dict_file) == 0)
    {
      printf("Cannot open dictionary file %s\n", dict_file);
      exit(0);
    }
  }
  else if (readdict(DICTFILE) == 0)
  {
    printf("Cannot open dictionary file %s\n", DICTFILE);
    exit(0);
  }

  /*
   * Ok - lets grab the expression now
   * (arbitrary decision - if expression on cmd line
   *  ignore the file argument)
   */
  if (radfilterfile)
  {
    int fd, nbytes = 1;
    char *bufptr, *envptr;

    /* We have a cmd filter */
    if (strcmp(radfilterfile, "-") == 0)
      fd = 0;
    else if (isalnum(*radfilterfile) && (envptr = getenv("RADSTOCK_PATH"))) {
      char tmpbuf[255];
      /* 
       * We'll check the RADSTOCK_PATH 
       */
      strcpy(radfilterbuf, envptr);
      bufptr = strtok(radfilterbuf, ":");
      while (bufptr && (fd < 0))
      {
	sprintf(tmpbuf, "%s/%s", bufptr, radfilterfile);
	if (debug) printf("Checking for file '%s'\n", tmpbuf);
	fd = open(tmpbuf, O_RDONLY);
	if (fd > 0)
	  printf("Config file found : %s\n", tmpbuf);
	else
	  bufptr = strtok(NULL, ":");
      }
    }
    else {
      fd = open(radfilterfile, O_RDONLY);
      if (fd > 0) printf("Config file found : %s\n", radfilterfile);
    }

    if (fd < 0)
    {
      printf("Cannot open config file\n");
      clean_exit(-1);
    }

    bufptr = radfilterbuf;
    while (nbytes)
    {
      nbytes =read(fd, (void *) bufptr, 1000);
      bufptr += nbytes;
    }
    *bufptr = '\0';
    if (fd) close(fd);
  }


  if (debug) printf("Handling radius filter '%s'\n", radfilterbuf);
  if (*radfilterbuf)
  {
    get_radfilter(radfilterbuf); 
    memset(watchlist, WATCHMAX * sizeof(struct t_watchlist), 0);
  }
  if (debug) printf("Done radius filter '%s'\n", radfilterbuf);

  //clean_exit(-2);

  if (read_file) {
    if (!(pd = pcap_open_offline(read_file, pc_err))) {
      perror(pc_err);
      clean_exit(-1);
    }

    live_read = 0;
    printf("input: %s\n", read_file);

  } else {
    if (!dev) {
      if (pcap_findalldevs(&alldevs,pc_err) == -1 ) { 
        perror(pc_err);
        clean_exit(-1);
      }
      // Traverse list of devices and use first one with a valid IP
      for ( d = alldevs; d; d = d->next ) {
        if ( pcap_lookupnet(d->name,&netp,&maskp,pc_err) == 0 ) {
          netp = ntohl(netp);
          if ( netp != 0x00000000 && netp != 0x7f000000 ) {
            // Use it
            dev = strdup(d->name);
            break;
          }
        }
      }
      pcap_freealldevs(alldevs);
    }
    if (!dev) {
      perror("No suitable interface found");
      clean_exit(-1);
    }
    if ((pd = pcap_open_live(dev, snaplen, promisc, to, pc_err)) == NULL) {
      perror(pc_err);
      clean_exit(-1);
    }

    if (pcap_lookupnet(dev, &net.s_addr, &mask.s_addr, pc_err) == -1) {
      perror(pc_err);
      memset(&net, 0, sizeof(net));
      memset(&mask, 0, sizeof(mask));
    } 

    if (!quiet) {
      printf("interface: %s", dev);
      if (net.s_addr && mask.s_addr) {
	printf(" (%s/", inet_ntoa(net));
	printf("%s)", inet_ntoa(mask)); 
      }
      printf("     ");
    }
  }


  if (argv) {
    filter = get_filter(&argv[optind]); 
    if (debug && !quiet) printf("filter1: %s\n", filter); 
    if (pcap_compile(pd, &pcapfilter, filter, 0, mask.s_addr)) {
      free(filter); 
      filter = get_filter(&argv[optind-1]); 
    if (debug && !quiet) printf("filter2: %s\n", filter); 

  //    pcap_restart();
      if (pcap_compile(pd, &pcapfilter, filter, 0, mask.s_addr)) {
	pcap_perror(pd, "pcap compile");
	clean_exit(-1);
      } else match_data = NULL;
    }

    if (!quiet) printf("filter: %s\n", filter); 
    
    if (pcap_setfilter(pd, &pcapfilter)) {
      pcap_perror(pd, "pcap set");
      clean_exit(-1);
    }
  }

  if (filter) free(filter);

  switch(pcap_datalink(pd)) {
  case DLT_EN10MB:
  case DLT_IEEE802:
    link_offset = ETHHDR_SIZE;
    break;

  case DLT_FDDI:
    link_offset = FDDIHDR_SIZE;
    break;

  case DLT_SLIP: 
    link_offset = SLIPHDR_SIZE;
    break;
    
  case DLT_PPP:
    link_offset = PPPHDR_SIZE;
    break;

  case DLT_RAW: 
    link_offset = RAWHDR_SIZE;
    break;

  case DLT_NULL:
    link_offset = LOOPHDR_SIZE;
    break;

  default:
    fprintf(stderr, "fatal: unsupported interface type\n");
    clean_exit(-1);
  }
  
  while (pcap_loop(pd, 0, (pcap_handler)process, 0));

  clean_exit(0);
  return 0;
}	


char *get_filter(char **argv) {
  char **arg = argv, *theirs, *mine;
  char *from, *to;
  int len = 0;

  if (!*arg)
  {
    len = strlen(defpcapfilter);
    if (!(mine = (char *) malloc(len+1)))
      return NULL;

    strcpy(mine, defpcapfilter);
    return mine;
  }

  while (*arg)
    len += strlen(*arg++) + 1;

  if (!(theirs = (char*)malloc(len + 1)) || 
      !(mine = (char*)malloc(len + sizeof(IP_ONLY))))
    return NULL;

  memset(theirs, 0, len + 1);
  memset(mine, 0, len + sizeof(IP_ONLY));

  arg = argv;
  to = theirs;
  while (*arg) 
  {
    from = *arg;
    while ((*to++ = *from++));
    *(to-1) = ' ';
    arg++;
  }

  sprintf(mine, IP_ONLY, theirs);

  free(theirs);
  return mine;
}


void clean_exit(int sig) {
  struct pcap_stat s;
  if (!quiet && sig >= 0) printf("exit\n");

  if (bin_data) free(bin_data);
  
  if (!quiet && sig >= 0 && !read_file &&
      pd && !pcap_stats(pd, &s)) 
    printf("%d received, %d dropped\n", s.ps_recv, s.ps_drop);

  if (pd) pcap_close(pd);

  exit(sig);
}


int strishex(char *str) {
  char *s;
  if ((s = strchr(str, 'x'))) 
    s++;
  else s = str;

  while (*s) 
    if (!isxdigit(*s++))
      return 0;

  return 1;
}



char *print_time(struct pcap_pkthdr *h) 
{
  static char timebuf[18];
  struct tm *t = localtime((time_t *) &h->ts.tv_sec);

  snprintf(timebuf,sizeof(timebuf), "[%02d:%02d:%02d.%06d]",
	  t->tm_hour, t->tm_min, t->tm_sec, h->ts.tv_usec);
  return timebuf;
}


char *print_time_date(struct pcap_pkthdr *h) 
{
  static char timebuf[29];
  struct tm *t = localtime((time_t *) &h->ts.tv_sec);

  snprintf(timebuf,sizeof(timebuf), "[%04d-%02d-%02d %02d:%02d:%02d.%06d]\n",
         t->tm_year+1900, t->tm_mon+1, t->tm_mday, t->tm_hour,
         t->tm_min, t->tm_sec, h->ts.tv_usec);
  return timebuf;
}


void print_time_diff(struct pcap_pkthdr *h, char *buf) { 
  unsigned secs, usecs;

  if (!prev_ts.tv_sec && !prev_ts.tv_usec) {// uninit
    prev_ts.tv_sec = h->ts.tv_sec;
    prev_ts.tv_usec = h->ts.tv_usec;
  }

  secs = h->ts.tv_sec - prev_ts.tv_sec;
  if (h->ts.tv_usec >= prev_ts.tv_usec)
    usecs = h->ts.tv_usec - prev_ts.tv_usec;
  else {
    secs--; 
    usecs = 1000000 - (prev_ts.tv_usec - h->ts.tv_usec);
  }

  printf("+%d.%06d ", secs, usecs);

  prev_ts.tv_sec = h->ts.tv_sec;
  prev_ts.tv_usec = h->ts.tv_usec;
}


void dump_delay(struct pcap_pkthdr *h) {
  unsigned secs, usecs;

  if (!prev_delay_ts.tv_sec && !prev_delay_ts.tv_usec) {
    prev_delay_ts.tv_sec = h->ts.tv_sec;
    prev_delay_ts.tv_usec = h->ts.tv_usec;
  }

  secs = h->ts.tv_sec - prev_delay_ts.tv_sec;
  if (h->ts.tv_usec >= prev_delay_ts.tv_usec)
    usecs = h->ts.tv_usec - prev_delay_ts.tv_usec;
  else {
    secs--; 
    usecs = 1000000 - (prev_delay_ts.tv_usec - h->ts.tv_usec);
  }

  sleep(secs);
  usleep(usecs);
  
  prev_delay_ts.tv_sec = h->ts.tv_sec;
  prev_delay_ts.tv_usec = h->ts.tv_usec;
}


void usage(int e) {
  printf("usage: radstock <options> <bpf filter>\n");
  printf("          -h      - display this page\n");
  printf("          -f file - define location of dictionary\n");
  printf("          -i intf - listen on given interface\n");
  printf("          -d lvl  - enable debugging output\n");
  printf("          -q      - quiet mode - do not show packet contents\n");
  printf("          -p      - run in non-promiscuous mode\n");
  printf("          -T      - print the time for each packet\n");
  printf("          -D      - print the date and time for each packet\n");
  printf("          -K      - only show authentication packets\n");
  printf("          -L      - only show accounting packets\n");
  printf("          -M      - only show request packets\n");
  printf("          -N      - only show response packets\n");
  printf("          -V      - show the version number\n");
  printf("          -e exp  - radius filter expression\n");
  printf("          -c file - read radius filter from file (- = stdin)\n");
  printf("\ni.e. radstock -e \"user-name = henry\" port 1645\n\n");

  exit(e);
}


void version(void) {
  printf("radstock: version %s\n", VERSIONSTR);
  exit(0);
}

